rclpy/src/rclpy/_rclpy_action.c

```diff
@@ -850,12 +1248,451 @@ rclpy_action_take_status(PyObject * Py_UNUSED(self), PyObject * args)
+  if (!action_server) {
+    return NULL;
+  }
+
+  destroy_ros_message_signature * destroy_cancel_request = NULL;
+  rcl_action_cancel_request_t * cancel_request =
+    (rcl_action_cancel_request_t *)rclpy_convert_from_py(pycancel_request, &destroy_cancel_request);
+  if (!cancel_request) {
+    return NULL;
+  }
+
+  rcl_action_cancel_response_t cancel_response = rcl_action_get_zero_initialized_cancel_response();
+  rcl_ret_t ret = rcl_action_process_cancel_request(
+    action_server, cancel_request, &cancel_response);
+  destroy_cancel_request(cancel_request);
+  if (RCL_RET_OK != ret) {
+    ret = rcl_action_cancel_response_fini(&cancel_response);
+    PyErr_Format(PyExc_RuntimeError,
+      "Failed to process cancel request: %s",
+      rcl_get_error_string().str);
+    rcl_reset_error();
+    return NULL;
+  }
+
+  PyObject * pycancel_response = rclpy_convert_to_py(&cancel_response.msg, pycancel_response_type);
+  ret = rcl_action_cancel_response_fini(&cancel_response);
+  if (!pycancel_response) {
+    return NULL;
+  }
+  if (RCL_RET_OK != ret) {
+    PyErr_Format(PyExc_RuntimeError,
+      "Failed to finalize cancel response: %s",
+      rcl_get_error_string().str);
+    rcl_reset_error();
+    return NULL;
+  }
+  return pycancel_response;
+}
+
+static PyObject *
+rclpy_action_expire_goals(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  PyObject * pyaction_server;
+  int64_t max_num_goals;
+
+  if (!PyArg_ParseTuple(args, "Ol", &pyaction_server, &max_num_goals)) {
+    return NULL;
+  }
+
+  rcl_action_server_t * action_server = (rcl_action_server_t *)PyCapsule_GetPointer(
+    pyaction_server, "rcl_action_server_t");
+  if (!action_server) {
+    return NULL;
+  }
+
+  rcl_action_goal_info_t * expired_goals =
+    (rcl_action_goal_info_t *)malloc(sizeof(rcl_action_goal_info_t) * max_num_goals);
+  if (!expired_goals) {
+    return PyErr_NoMemory();
+  }
+  size_t num_expired;
+  rcl_ret_t ret = rcl_action_expire_goals(
+    action_server, expired_goals, max_num_goals, &num_expired);
+  if (RCL_RET_OK != ret) {
+    PyErr_Format(PyExc_RuntimeError, "Failed to expire goals: %s", rcl_get_error_string().str);
+    rcl_reset_error();
+    free(expired_goals);
+    return NULL;
+  }
+
+  // Get Python GoalInfo type
+  PyObject * pyaction_msgs_module = PyImport_ImportModule("action_msgs.msg");
+  if (!pyaction_msgs_module) {
+    free(expired_goals);
+    return NULL;
+  }
+  PyObject * pygoal_info_class = PyObject_GetAttrString(pyaction_msgs_module, "GoalInfo");
+  Py_DECREF(pyaction_msgs_module);
+  if (!pygoal_info_class) {
+    free(expired_goals);
+    return NULL;
+  }
+  PyObject * pygoal_info_type = PyObject_CallObject(pygoal_info_class, NULL);
+  Py_DECREF(pygoal_info_class);
+  if (!pygoal_info_type) {
+    free(expired_goals);
+    return NULL;
+  }
+
+  // Create a tuple of GoalInfo instances to return
+  PyObject * result_tuple = PyTuple_New(num_expired);
+  if (!result_tuple) {
+    free(expired_goals);
+    return NULL;
```

Comment:
`Py_DECREF(pygoal_info_type)` before returning

Reply:
Nice catch. Fixed in b037186